algebra/default/vector.c

```diff
@@ -0,0 +1,789 @@
+  c_float* uv = u->values;
+
+  for (i = 0; i < length; i++) {
+    if (uv[i]   > +infval) {       // Infinite upper bound
+      if (lv[i] < -infval) {       // Infinite lower bound
+        // Both bounds infinite
+        yv[i] = 0.0;
+      } else {
+        // Only upper bound infinite
+        yv[i] = c_min(yv[i], 0.0);
+      }
+    } else if (lv[i] < -infval) {  // Infinite lower bound
+      // Only lower bound infinite
+      yv[i] = c_max(yv[i], 0.0);
+    }
+  }
+}
+
+c_int OSQPVectorf_in_polar_reccone(const OSQPVectorf *y,
+                                   const OSQPVectorf *l,
+                                   const OSQPVectorf *u,
+                                   c_float           infval,
+                                   c_float           tol){
+
+  c_int i; // Index for loops
+
+  c_int    length = y->length;
+  c_float* yv     = y->values;
+  c_float* lv     = l->values;
+  c_float* uv     = u->values;
+
+  for (i = 0; i < length; i++) {
+    if (((uv[i] < +infval) &&
+         (yv[i] > +tol)) ||
+        ((lv[i] > -infval) &&
+         (yv[i] < -tol))) {
+      // At least one condition not satisfied -> not dual infeasible
+      return 0;
+    }
+  }
+  return 1;
+}
+
+
+void OSQPVectorf_permute(OSQPVectorf *x, const OSQPVectorf *b, const OSQPVectori *p){
+
+  c_int j;
+  c_int length = x->length;
+  c_float*  xv = x->values;
+  c_float*  bv = b->values;
+  c_int*    pv = p->values;
+
+  for (j = 0; j < length; j++) {
+    xv[j] = bv[pv[j]];
+  }
+}
+
+void OSQPVectori_permute(OSQPVectori *x, const OSQPVectori *b, const OSQPVectori *p){
+
+  c_int j;
+  c_int length = x->length;
+  c_int*    xv = x->values;
+  c_int*    bv = b->values;
+  c_int*    pv = p->values;
+
+  for (j = 0; j < length; j++) {
+    xv[j] = bv[pv[j]];
+  }
+}
+
+void OSQPVectorf_ipermute(OSQPVectorf *x, const OSQPVectorf *b, const OSQPVectori *p){
+
+  c_int j;
+  c_int length = x->length;
+  c_float*  xv = x->values;
+  c_float*  bv = b->values;
+  c_int*    pv = p->values;
+
+  for (j = 0; j < length; j++) {
+    xv[pv[j]] = bv[j];
+  }
+}
+
+void OSQPVectori_ipermute(OSQPVectori *x, const OSQPVectori *b, const OSQPVectori *p){
+
+  c_int j;
+  c_int length = x->length;
+  c_int*    xv = x->values;
+  c_int*    bv = b->values;
+  c_int*    pv = p->values;
+
+  for (j = 0; j < length; j++) {
+    xv[pv[j]] = bv[j];
+  }
+}
+
+
+
+#if EMBEDDED != 1
+
+c_float OSQPVectorf_mean(const OSQPVectorf *a){
+
+  if (a->length) {
+    return OSQPVectorf_sum(a) / (a->length);
+  }
+  else {
+    return 0;
+  }
+}
+
+void OSQPVectorf_ew_reciprocal(OSQPVectorf *b, const OSQPVectorf *a){
+
+  c_int i;
+  c_int length = a->length;
+  c_float*  av = a->values;
+  c_float*  bv = b->values;
+
+  for (i = 0; i < length; i++) {
+    bv[i] = (c_float)1.0 / av[i];
+  }
+}
+
+void OSQPVectorf_ew_sqrt(OSQPVectorf *a){
+
+  c_int i;
+  c_int length = a->length;
+  c_float*  av = a->values;
+
+  for (i = 0; i < length; i++) {
+    av[i] = c_sqrt(av[i]);
+  }
+}
+
```

Comment:
It was unclear to me whether the purpose of this logic in the previous code (i.e. the equivalent operations on c_float arrays) was meant to avoid division or to exclude compilation of all functions that are not required for refactorisation.    

If the purpose of EMBEDDED=1 is only to avoid division, then we should exclude only the mean and ew division functions (and the square root? - it is not clear to me).      If it is to do that plus trying to minimize the compiled footprint, then it should stay as it is.

It is unclear to me if we can rely generically on all compilers to drop unused functions when linking.   It is also unclear to me if we care since these functions are all so small.

Reply:
Noted as a v1.0 update [here](https://github.com/oxfordcontrol/osqp/projects/1#column-1967224).